crates/handlers/src/views/login.rs

```diff
@@ -207,76 +212,116 @@ pub(crate) async fn post(
+                    .map_err(|_err| FormError::Internal)?;
+
+                // Replicate in Synapse
+                homeserver
+                    .provision_user(&ProvisionRequest::new(mxid.clone(), username.to_owned()))
+                    .await
+                    .unwrap();
+
+                new_user
+            };
+            // Update the password if needed
+            let result = password_manager
+                .hash(&mut rng, password.to_owned().into_bytes().into())
+                .await;
+            let (version, hashed_password) = match result {
+                Ok((version, hashed_password)) => (version, hashed_password),
+                Err(_err) => return Err(FormError::Internal),
+            };
+            repo.user_password()
+                .upsert(&mut rng, clock, &user, version, hashed_password)
+                .await
+                .map_err(|_err| FormError::Internal)?;
+
+            user
+        } else {
+            return Err(FormError::InvalidCredentials);
+        }
+    } else {
+        // If rest_auth_provider is not enabled, proceed with the normal authentication
+        user_login_with_password(
             &mut rng,
-            user_password.version,
-            password,
-            user_password.hashed_password.clone(),
+            clock,
+            &password_manager,
+            &limiter,
+            requester,
+            repo,
+            username.to_owned(),
+            password.to_owned(),
         )
         .await
-        .map_err(|_| FormError::InvalidCredentials)?;
-
-    let user_password = if let Some((version, new_password_hash)) = new_password_hash {
-        // Save the upgraded password
-        repo.user_password()
-            .add(
-                &mut rng,
-                clock,
-                &user,
-                version,
-                new_password_hash,
-                Some(&user_password),
-            )
-            .await
-            .map_err(|_| FormError::Internal)?
-    } else {
-        user_password
+        .map_err(|err| match err {
+            RouteError::RateLimited(_) => FormError::RateLimitExceeded,
+            RouteError::PasswordVerificationFailed(_)
+            | RouteError::UserNotFound
+            | RouteError::NoPassword => FormError::InvalidCredentials,
+            _ => FormError::Internal,
+        })?
     };
 
+    // Start a new compat session without verifying the password again
+    start_new_session(&mut rng, clock, repo, user.clone(), None)
+        .await
+        .map_err(|_err| FormError::Internal)?;
     // Start a new session
     let user_session = repo
         .browser_session()
         .add(&mut rng, clock, &user, user_agent)
         .await
         .map_err(|_| FormError::Internal)?;
 
+    let user_password = repo
+        .user_password()
+        .active(&user)
+        .await
+        .map_err(|_| FormError::InvalidCredentials)?
+        .ok_or(FormError::InvalidCredentials)?;
```

Comment:
Oh, is that why? To get an authentication session going on?
I feel like this should get recorded another way, maybe a new table recording authentications from an external password verification service?

Reply:
Yes exactly I wanted to have an actual session going on after the call to the login API is made, so subsequent whoami calls and things like that actually work. If there is a risk to mess things up with having every compat sessions on the same table, we can store them in another table, but if it's not that risky it may be a bit overkill. In Synapse if you have REST password enabled all login requests go to the REST password provider, and it's the same with this implementation, if REST password provider is configured every compat login goes it. This is why it seemed logical to just use the same table, since when this config is enabled a compat session is unambiguously always a REST password provider compat session.